spherical_splines/SHTOOLS/man/man1/shlocalizedadmitcorr.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SHLOCALIZEDADMITCORR 1"
.TH SHLOCALIZEDADMITCORR 1 "2009-08-18" "SHTOOLS 2.6" "SHTOOLS 2.6"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "SHLocalizedAdmitCorr"
.IX Header "SHLocalizedAdmitCorr"
.IP "SHLocalizedAdmitCorr \-" 23
.IX Item "SHLocalizedAdmitCorr -"
Calculate the localized admittance and correlation spectra of two functions at a given location.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 SHLocalizedAdmitCorr (" 34
.IX Item "SUBROUTINE SHLocalizedAdmitCorr ("
\&\s-1TAPERS\s0, \s-1TAPER_ORDER\s0, \s-1LWIN\s0, \s-1LAT\s0, \s-1LON\s0, \s-1GILM\s0, \s-1TILM\s0, \s-1LMAX\s0, \s-1ADMIT\s0, \s-1CORR\s0, K, \s-1ADMIT_ERROR\s0, \s-1CORR_ERROR\s0, \s-1TAPER_WT\s0, \s-1MTDEF\s0, K1LINSIG )
.RS 4
.IP "REAL*8" 19
.IX Item "REAL*8"
\&\s-1TAPERS\s0(\s-1LWIN+1\s0, K), \s-1LAT\s0, \s-1LON\s0, \s-1GILM\s0(2,\ \s-1LMAX+1\s0,\ \s-1LMAX+1\s0), \s-1TILM\s0(2,\ \s-1LMAX+1\s0,\ \s-1LMAX+1\s0), \s-1ADMIT\s0(\s-1LMAX\-LWIN+1\s0), \s-1CORR\s0(\s-1LMAX\-LWIN+1\s0)
.IP "\s-1INTEGER\s0" 19
.IX Item "INTEGER"
\&\s-1TAPER_ORDER\s0(K), \s-1LWIN\s0, \s-1LMAX\s0, K
.IP "REAL*8, \s-1OPTIONAL\s0" 19
.IX Item "REAL*8, OPTIONAL"
\&\s-1ADMIT_ERROR\s0(\s-1LMAX\-LWIN+1\s0), \s-1CORR_ERRR\s0(\s-1LMAX\-LWIN+1\s0), \s-1TAPER_WT\s0(K)
.IP "\s-1INTEGER\s0, \s-1OPTIONAL\s0" 19
.IX Item "INTEGER, OPTIONAL"
\&\s-1MTDEF\s0, K1LINSIG
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fISHLocalizedAdmitCorr\fR will calculate the localized admittance and degree correlation spectra of two functions at a given location. The windowing functions are assumed to be solutions to the spherical-cap concentration problem (as calculated by \fISHReturnTapers\fR or \fISHReturnTapersM\fR), of which the best K concentrated tapers are utilized. If K is greater than 1, then estimates of the standard error for the admittance and correlation will be returned in the optional arrays \s-1ADMIT_ERROR\s0 and \s-1CORR_ERROR\s0. The symmetry axis of the localizing windows are rotated to the coordinates (\s-1LAT\s0, \s-1LON\s0) before performing the windowing operation.
.PP
The admittance is defined as Sgt/Stt, where Sgt is the localized cross-power spectrum of two functions G and T expressed in spherical harmonics. The localized degree-correlation spectrum is defined as Sgt / sqrt( Sgg Stt ), which can possess values between \-1 and 1. Two methods are available for calculating the multitaper admittance and correlation functions. When \s-1MTDEF\s0 is 1 (default), the multitaper estimates and errors of Sgt, Stt, and Sgg are calculated by calls to \fISHMultiTaperSE\fR and \fISHMultiTaperCSE\fR, and these results are then used to calculate the final admittance and correlation functions. When \s-1MTDEF\s0 is 2, the admitance and correlation are calculated invidivually for each individual taper, and these results are then averaged.
.PP
If the optional parameter K1LINSIG is specified, and only a single taper is being used, the uncertainty in the admittance function will be calculated by assuming the gravity and topography coefficients are related by a linear degree-dependent transfer function and that the lack of correlation is a result of uncorrelated noise. These errors should not be used if it is suspected that the gravity and topography are the result of partially to unccorrelated surface  and subsurface loads.
.PP
When \s-1MTDEF\s0 is 1, by default, the multitaper spectral estimates are calculated as an unweighted average of the individual tapered estimates. However, if the optional argument \s-1TAPER_WT\s0 is specified, a weighted average will be employed using the weights in this array. Minimum variance optimal weights can be obtained from the routines \fISHMTVarOpt\fR and \fISHMTVarOpt0\fR if the form of the underlying global power spectrum is known (or can be estimated). Taper weights can not be used when \s-1MTDEF\s0 is 2
.PP
This routine assumes that the input functions and tapers are expressed using geodesy 4\-pi normalized spherical harmonic functions that exclude the  Condon-Shortley phase factor of (\-1)^m.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "\s-1TAPERS\s0" 14
.IX Item "TAPERS"
(input) REAL*8, \s-1DIMENSION\s0 (\s-1LWIN+1\s0, K)
.Sp
A matrix of localization functions obtained from \fISHReturnTapers\fR or \fISHReturnTapersM\fR.
.IP "\s-1TAPER_ORDER\s0" 14
.IX Item "TAPER_ORDER"
(input) \s-1INTEGER\s0, \s-1DIMENSION\s0 (K)
.Sp
The angular order of the windowing coefficients in \s-1TAPERS\s0. If this matrix was created using \fISHReturnTapersM\fR, then this array must be composed of zeros.
.IP "\s-1LWIN\s0" 14
.IX Item "LWIN"
(input) \s-1INTEGER\s0
.Sp
The spherical harmonic bandwidth of the localizing windows.
.IP "\s-1LAT\s0" 14
.IX Item "LAT"
(input) REAL*8
.Sp
The latitude of the localized analysis in \s-1DEGREES\s0.
.IP "\s-1LON\s0" 14
.IX Item "LON"
(input) REAL*8
.Sp
The longitude of the localized analysis in \s-1DEGREES\s0.
.IP "\s-1GILM\s0" 14
.IX Item "GILM"
(input) REAL*8, \s-1DIMENSION\s0 (2, \s-1LMAX+1\s0, \s-1LMAX+1\s0)
.Sp
The spherical harmonic coefficients of the function G.
.IP "\s-1TILM\s0" 14
.IX Item "TILM"
(input) REAL*8, \s-1DIMENSION\s0 (2, \s-1LMAX+1\s0, \s-1LMAX+1\s0)
.Sp
The spherical harmonic coefficients of the function T.
.IP "\s-1LMAX\s0" 14
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum spherical harmonic degree of the input functions corresponding to \s-1GILM\s0 and \s-1TILM\s0.
.IP "\s-1ADMIT\s0" 14
.IX Item "ADMIT"
(output) REAL*8, \s-1DIMENSION\s0 (\s-1LMAX\-LWIN+1\s0)
.Sp
The admittance function, which is equal to Sgt / Stt.
.IP "\s-1CORR\s0" 14
.IX Item "CORR"
(output) REAL*8, \s-1DIMENSION\s0 (\s-1LMAX\-LWIN+1\s0)
.Sp
The degree correlation function, which is equal to Sgt / sqrt ( Sgg Stt ).
.IP "K" 14
.IX Item "K"
(input) \s-1INTEGER\s0
.Sp
The number of tapers to be used in the multitaper spectral analysis.
.IP "\s-1ADMIT_ERROR\s0" 14
.IX Item "ADMIT_ERROR"
(output) REAL*8, \s-1OPTIONAL\s0, \s-1DIMENSION\s0 (\s-1LMAX\-LWIN+1\s0)
.Sp
The standard error of the admittance function.
.IP "\s-1CORR_ERROR\s0" 14
.IX Item "CORR_ERROR"
(output) REAL*8, \s-1OPTIONAL\s0, \s-1DIMENSION\s0 (\s-1LMAX\-LWIN+1\s0)
.Sp
The standard error of the degree correlation function.
.IP "\s-1TAPER_WT\s0" 14
.IX Item "TAPER_WT"
(input) REAL*8, \s-1OPTIONAL\s0, \s-1DIMENSION\s0 (K)
.Sp
The weights to be applied to the spectral estimates when calculating the admittance, correlation, and their associated errors. This must sum to unity.
.IP "\s-1MTDEF\s0" 14
.IX Item "MTDEF"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
1 (default): Calculate the multitaper spectral estimates Sgt, Sgg and Stt first, and then use these to calculate the admittance and correlation functions. 2: Calculate admittance and correlation spectra using each individual taper, and then average these to obtain the multitaper admittance and correlation functions.
.IP "K1LINSIG" 14
.IX Item "K1LINSIG"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
If equal to one, and only a single taper is being used, the errors in the admittance function will be calculated by assuming that the gravity and topography coefficients are related by a linear degree-dependent transfer function and that the lack of correlation is a result of uncorrelated noise. These errors should not be used if it is suspected that the gravity and topography are the result of partially correlated loads. This is the square root of eq. 33 of Simons et al. 1997.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIshreturntapers\fR\|(1), \fIshreturntapersm\fR\|(1), \fIshmultitaperse\fR\|(1), \fIshmultitapercse\fR\|(1)
.PP
<http://www.ipgp.jussieu.fr/~wieczor/SHTOOLS/SHTOOLS.html>
.SH "REFERENCES"
.IX Header "REFERENCES"
Wieczorek, M. A. and F. J. Minimum-variance multitaper spectral estimation on the sphere, \fIJ. Fourier Anal. Appl.\fR, 13, doi:10.1007/s00041\-006\-6904\-1, 665\-692, 2007.
.PP
Simons, F. J., F. A. Dahlen and M. A. Wieczorek, Spatiospectral concentration on the sphere, \fI\s-1SIAM\s0 Review\fR, 48, 504\-536, doi:10.1137/S0036144504445765, 2006.
.PP
Wieczorek, M. A. and F. J. Simons, Localized spectral analysis on the sphere, 
\&\fIGeophys. J. Int.\fR, 162, 655\-675, 2005.
.PP
Simons, M., S. C. Solomon and B. H. Hager, Localization of gravity and topography: constrains on the tectonics and mantle dynamics of Venus, \fIGeophys. J. Int.\fR, 131, 24\-44, 1997.
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2005\-2009 by Mark Wieczorek <wieczor@ipgp.fr>.
.PP
This is free software; you can distribute and modify it under the terms of the revised \s-1BSD\s0 license.


spherical_splines/SHTOOLS/src/SHLocalizedAdmitCorr.f95
subroutine SHLocalizedAdmitCorr(tapers, taper_order, lwin, lat, lon, g, t, lmax, admit, corr, K, &
	admit_error, corr_error, taper_wt, mtdef, k1linsig)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	Given two spherical harmonic fields (G and T), this routine will calculate
!	the localized admittance and correlation using the first space-concentrated
!	window of Wieczorek and Simons (2005). All functions must be 4-pi normalized, 
!	and exclude the Condon-Shortley phase factor. Two manners of calculating the 
!	localized admittance and correlation are possible according to the optional 
!	parameter MTDEF. In one case, the multitaper cross-power spectra are calculated,
!	and from these, the admittance and correlation. In the second, the admittance and 
!	correlation are calculated for each taper, and these are then averaged.
!
!	Calling Parameters
!		IN
!			tapers		A matrix of tapers obtained from SHReturnTapers.
!			taper_order	A vector continaing the angular order of each column of taper.
!			lwin		Spectral bandwidth of the localizing window.
!			lat, lon	Latitude and longitude that the window will be
!					rotated to, in DEGREES.
!			G, T		Input spherical harmonic fields.
!			K		Number of tapers to use in Multitaper spectral estimations.
!			lmax		Maximum spherical harmonic degree of the intput fields.
!		OUT
!			admit		Admittance between the localized G and T assuming 
!					that G = Z T.
!			
!			corr		Correlation of the two fields.
!		OPTIONAL (OUT)
!			admit_error	Error of the admittance (only when K>1)
!			corr_error	Error of the admittance (only when K>1)
!		OPTIONAL (IN)
!			mtdef		1 (default): Calculate multitaper cross-spectral estimates, and use these
!					to calculate a single admittance and correlations.
!					2: Calculate the admittance and correlation using each individual taper, and
!					then average these to get the admittance and correlation.
!			taper_wt	Weights to be applied to the spectral estimates. This can only be used
!					when MTDEF is 1.
!			k1linsig:	If present and equal to 1, the uncertainty in the addmittance
!					will be calculated by assuming the gravity and topography coefficients
!					are linearly correlated, and that any lack of correlation is the 
!					result of uncorrelated noise. This should only be used when one expects
!					the gravity and topography to be linearly correlated and when only a 
!					single taper is being used. This should not be used with a Forsyth 
!					type model that predicts a less than 1correlation coefficient. 
!					This is the square root of eq. 33 of Simons et al. 1997.
!
!
!	Notes:
!		1. The units of the output admittance will correspond to the units of the
!		spherical harmonic coefficients. If gravity/topography admittances are
!		desired, then either the gravity coefficients should be multiplied by
!		G M (l+1) * (r0/r)**(l+2) / r0**2 before calling this routine, or
!		the admittances should be multiplied by this factor afterwards.
!
!		2. The correlation is defined as Sgt / sqrt(Sgg Stt), which varies between -1 and 1.
!		To obtain the "coherence" (which is also sometimes referred to as the "coherence squared"),
!		just square this number.
!
!	Dependencies: 	SHMultitaperSE, SHMultiTaperCSE, CSPHASE_DEFAULT,  djpi2, SHRotateRealCoef,
!			SHMultiply, SHCrossPowerSpectrum, SHPowerSpectrum
!
!	Written by Mark Wieczorek July 2004
!		Completely rewritten June 1 2006 to take into account non-zonal windows.
!		April 2009: added the optional parameter MTDEF for defining two manners of calculating
!			mutlitapered admittances and correlations.
!
!	Copyright (c) 2009, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	use SHTOOLS, only: SHMultiTaperSE, SHMultiTaperCSE, djpi2, SHRotateRealCoef, &
				SHMultiply, SHCrossPowerSpectrum, SHPowerSpectrum
			
	implicit none

	real*8, intent(in) ::	tapers(:,:), lat, lon, g(:,:,:), t(:,:,:)
	integer, intent(in) ::	lwin, lmax, K, taper_order(:)
	real*8, intent(out) ::	admit(:), corr(:)
	real*8, intent(out), optional ::	admit_error(:), corr_error(:)
	integer, intent(in), optional ::	mtdef, k1linsig
	real*8, intent(in), optional ::	taper_wt(:)
	integer ::		lmaxwin, l, def, astat(5), phase, norm, i
	real*8 ::		pi, g_power(2,lwin+lmax+1), t_power(2,lwin+lmax+1), gt_power(2,lwin+lmax+1), x(3), &
				sgt(lmax-lwin+1, K), sgg(lmax-lwin+1, K), stt(lmax-lwin+1, K), admit_k(lmax-lwin+1, K), &
				corr_k(lmax-lwin+1, K)
	real*8, allocatable ::	dj(:,:,:), shwin(:,:,:), shwinrot(:,:,:), shloc_g(:,:,:), shloc_t(:,:,:)	

	phase = 1
	norm = 1

	pi = acos(-1.0d0)
	lmaxwin = lmax + lwin

	if (present(k1linsig) .and. K /= 1) then
		if (k1linsig == 1) then
			print*, "SHlocalizedAdmitCorr --- Error"
			print*, "If K1LINSIG is present and equal to 1, K must be equal to 1."
			print*, "Input value of K1LINSIG is ", k1linsig
			stop
		endif
	endif

	if (size(admit) < lmax-lwin+1) then
		print*, "Error --- SHLocalizedAdmitCorr"
		print*, "ADMIT must be dimensioned as (LMAX-LWIN+1) where LMAX and LWIN are ", lmax, lwin
		print*, "Input array is dimensioned ", size(admit)
		stop
	elseif (size(corr) < lmax-lwin+1) then
		print*, "Error --- SHLocalizedAdmitCorr"
		print*, "CORR must be dimensioned as (LMAX-LWIN+1) where LMAX and LWIN are ", lmax, lwin
		print*, "Input array is dimensioned ", size(corr)
		stop
	elseif (size(tapers(:,1)) < lwin+1 .or. size(tapers(1,:)) < K) then
		print*, "Error --- SHLocalizedAdmitCorr"
		print*, "TAPERS must be dimensioned as (LWIN+1, K) where lwin and K are ", lwin, K
		print*, "Iinput array is dimensioned as ", size(tapers(:,1)), size(tapers(1,:))
		stop
	elseif (size(taper_order) < K ) then
		print*, "Error --- SHLocalizedAdmitCorr"
		print*, "TAPER_ORDER must be dimensioned as (K) where K is ", K
		print*, "Input array is dimensioned ", size(taper_order)
		stop
	elseif (size(g(:,1,1)) < 2 .or. size(g(1,:,1)) < lmax+1 .or. size(g(1,1,:)) < lmax+1) then
		print*, "Error --- SHLocalizedAdmitCorr"
		print*, "G must be dimensioned as (2, LMAX+1, LMAX+1) where lmax is ", lmax
		print*, "Input array is dimensioned ", size(g(:,1,1)), size(g(1,:,1)), size(g(1,1,:))
		stop
	elseif (size(t(:,1,1)) < 2 .or. size(t(1,:,1)) < lmax+1 .or. size(t(1,1,:)) < lmax+1) then
		print*, "Error --- SHLocalizedAdmitCorr"
		print*, "T must be dimensioned as (2, LMAX+1, LMAX+1) where lmax is ", lmax
		print*, "Input array is dimensioned ", size(t(:,1,1)), size(t(1,:,1)), size(t(1,1,:))
		stop
	endif
	
	if (present(admit_error)) then
		if (size(admit_error) < lmax - lwin+1) then
			print*, "Error ---SHLocalizedAdmitCorr"
			print*, "ADMIT_ERROR must be dimensioned as (LMAX-LWIN+1) where lmax and lmaxt are ", lmax, lwin
			print*, "Input array is dimensioned ", size(admit_error)
			stop	
		endif
	endif
	
	if (present(corr_error)) then
		if (size(corr_error) < lmax - lwin + 1) then
			print*, "Error ---SHLocalizedAdmitCorr"
			print*, "CORR_ERROR  must be dimensioned as (LMAX-LWIN+1) where lmax and lmaxt are ", lmax, lwin
			print*, "Input array is dimensioedn ", size(corr_error)
			stop	
		endif	
	endif
	
	if(present(taper_wt)) then
		if (size(taper_wt) < K) then
			print*, "Error --- SHLocalizedAdmitCorr"
			print*, "TAPER_WT must be dimensioned as (K) where K is ", K
			print*, "Input array has dimension ", size(taper_wt)
			stop
		endif
	endif
	
	if (present(mtdef)) then
		if (mtdef == 2 .and. present(taper_wt)) then
			print*, "Error --- SHLocalizedAdmitCorr"
			print*, "TAPER_WT can only be used when MTDEF is 1."
			stop
		endif
	endif
	
	if (present(mtdef)) then
		if (mtdef /= 1 .and. mtdef /= 2) then
			print*, "SHLocalizedAdmitCorr --- Error"
     			print*, "MTDEF must be 1 or 2."
     			print*, "Input value is ", mtdef
     			stop
     		else
     			def = mtdef
     		endif
     	else
     		def = 1
     	endif
     	
     	
   	admit = 0.0d0
   	corr = 0.0d0
   	
     	if (present(admit_error)) then
		admit_error = 0.0d0
	endif
	
	if (present(corr_error)) then
		corr_error = 0.0d0
	endif

	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Determine multitaper cross spectra estimates of G and T, and then calculate 
	!	the admittance and correlation. Errors for the latter are calculated by adding
	!	the error sources in quadrature. Taper weights can be specified in order
	!	to minimize the variance of the multitaper cross-spectral estimates.
	!
	!	Note that only the admittances with degrees greater than lwin and less than
	!	Lmax - lwin should be interpretted.
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		
	if (def == 1) then
	
		if (present(taper_wt)) then

			call SHMultiTaperSE(g_power(1,:), g_power(2,:), g, lmax, tapers(1:lwin+1,1:K), &
				taper_order(1:K), lwin, K, lat=lat, lon=lon, taper_wt = taper_wt(1:K), csphase = phase)
			call SHMultiTaperSE(t_power(1,:), t_power(2,:), t, lmax, tapers(1:lwin+1,1:K), &
				taper_order(1:K), lwin, K, lat=lat, lon=lon, taper_wt = taper_wt(1:K), csphase = phase)
			call SHMultiTaperCSE(gt_power(1,:), gt_power(2,:), g, lmax, t, lmax, tapers(1:lwin+1,1:K), &
				taper_order(1:K), lwin, K, lat=lat, lon=lon, taper_wt = taper_wt(1:K), csphase = phase)
		
		else
	
			call SHMultiTaperSE(g_power(1,:), g_power(2,:), g, lmax, tapers(1:lwin+1,1:K), &
				taper_order(1:K), lwin, K, lat=lat, lon=lon, csphase = phase)
			call SHMultiTaperSE(t_power(1,:), t_power(2,:), t, lmax, tapers(1:lwin+1,1:K), &
				taper_order(1:K), lwin, K, lat=lat, lon=lon, csphase = phase)
			call SHMultiTaperCSE(gt_power(1,:), gt_power(2,:), g, lmax, t, lmax, tapers(1:lwin+1,1:K), &
				taper_order(1:K), lwin, K, lat=lat, lon=lon, csphase = phase)
		endif
	
		admit(1:lmax-lwin+1) = gt_power(1,1:lmax-lwin+1) / t_power(1,1:lmax-lwin+1)
		corr(1:lmax-lwin+1) =  gt_power(1,1:lmax-lwin+1) / sqrt(t_power(1,1:lmax-lwin+1)*g_power(1,1:lmax-lwin+1))

		if (K > 1 .and. present(admit_error)) then
			admit_error(1:lmax-lwin+1) = ( gt_power(2,1:lmax-lwin+1) / t_power(1,1:lmax-lwin+1) )**2 + &
				( gt_power(1,1:lmax-lwin+1) / t_power(1,1:lmax-lwin+1)**2 * t_power(2,1:lmax-lwin+1) )**2
			admit_error(1:lmax-lwin+1) = sqrt(admit_error(1:lmax-lwin+1))
		endif
	
		if (K > 1 .and. present(corr_error)) then
			corr_error(1:lmax-lwin+1) = gt_power(2,1:lmax-lwin+1)**2 / t_power(1,1:lmax-lwin+1) / g_power(1,1:lmax-lwin+1) + &
				( gt_power(1,1:lwin-lmax+1) * t_power(2,1:lwin-lmax+1) / sqrt(g_power(1,1:lmax-lwin+1)) &
				/ 2.0d0 / t_power(1,1:lmax-lwin+1)**(3.0d0/2.0d0) )**2 + &
				( gt_power(1,1:lwin-lmax+1) * g_power(2,1:lwin-lmax+1) / sqrt(t_power(1,1:lmax-lwin+1)) &
				/ 2.0d0 / g_power(1,1:lmax-lwin+1)**(3.0d0/2.0d0) )**2
			corr_error(1:lmax-lwin+1) = sqrt(corr_error(1:lmax-lwin+1))
		endif
	
		if (K==1 .and. present(k1linsig) .and. present(admit_error)) then
			admit_error = 0.0d0
			if (k1linsig == 1) then
				do l=1, lmax-lwin
					admit_error(l+1) = g_power(1,l+1)*(1.0d0 - corr(l+1)**2) / ( t_power(1,l+1) * dble(2*l) )
				enddo
				admit_error(1:lmax-lwin+1) = sqrt(admit_error(1:lmax-lwin+1))
			endif
		endif
		
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Calculate the admittance and correlation for each individual taper. Then average
	!	these in order to get the multitaper estimates and uncertainties.
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

	else
		
		x(1) = 0.0d0
		x(2) = -(90.0d0 - lat)*pi/180.0d0
		x(3) =  -lon*pi/180.0d0
		
		allocate(dj(lwin+1,lwin+1,lwin+1), stat = astat(1))
		allocate(shwin(2,lwin+1,lwin+1), stat = astat(2))
		allocate(shwinrot(2,lwin+1,lwin+1), stat = astat(3))
		allocate(shloc_g(2, lmaxwin+1, lmaxwin+1), stat= astat(4))
		allocate(shloc_t(2, lmaxwin+1, lmaxwin+1), stat= astat(5))
		
		if (astat(1) /= 0 .or. astat(2) /= 0 .or. astat(3) /= 0 .or. astat(4) /= 0 .or. astat(5) /= 0) then
			print*, "Error --- SHLocalizedAdmitCorr"
			print*, "Problem allocating arrays DJ, SHWIN, SHWINROT, SHLOC_G, SHLOC_T", &
				astat(1), astat(2), astat(3), astat(4), astat(5)
			stop
		endif
		
		dj = 0.0d0
		shwinrot = 0.0d0
		shloc_g = 0.0d0
		shloc_t = 0.0d0
		
		call djpi2(dj, lwin)	! Create rotation matrix used in the rotation routine.
		
		do i=1, K
			shwin = 0.0d0
			if (taper_order(i) < 0) then
				shwin(2,1:lwin+1,abs(taper_order(i))+1) = tapers(1:lwin+1,i)
			else
				shwin(1,1:lwin+1,taper_order(i)+1) = tapers(1:lwin+1,i)
			endif
		
			call SHRotateRealCoef(shwinrot, shwin, lwin, x, dj)
		
			call SHMultiply(shloc_g, g, lmax, shwinrot, lwin, csphase = phase, norm=norm)
			call SHMultiply(shloc_t, t, lmax, shwinrot, lwin, csphase = phase, norm=norm)
			
			call SHCrossPowerSpectrum(shloc_g, shloc_t, lmax-lwin, sgt(:,i))
			call SHPowerSpectrum(shloc_g, lmax-lwin, sgg(:,i))
			call SHPowerSpectrum(shloc_t, lmax-lwin, stt(:,i))
			
			admit_k(1:lmax-lwin+1, i) =  sgt(1:lmax-lwin+1, i) / stt(1:lmax-lwin+1, i)
			corr_k(1:lmax-lwin+1, i) =  sgt(1:lmax-lwin+1, i) / sqrt(stt(1:lmax-lwin+1, i)) / sqrt(sgg(1:lmax-lwin+1, i))
			
		enddo
		
		do l=0, lmax-lwin, 1
			admit(l+1) = sum(admit_k(l+1,1:K)) / dble(K)
			corr(l+1) = sum(corr_k(l+1,1:K)) / dble(K)
		enddo
		
		if (present(admit_error) .or. present(corr_error)) then
		
			do l=0, lmax-lwin, 1
				
				if (present(admit_error)) then
				
					if (K > 1) then 
					
						admit_error(l+1) = sum( ( admit_k(l+1,1:K) - admit(l+1) )**2 ) / dble(K-1) / dble(K) ! standard error!
						admit_error(l+1) = sqrt(admit_error(l+1))

					elseif (K == 1 .and. present(k1linsig)) then
					
						if (k1linsig == 1) then
						
							if (l==0) then
								admit_error(1) = 0.0d0
							else
								admit_error(l+1) = sgg(l+1,1)*(1.0d0 - corr(l+1)**2) / ( stt(l+1,1) * dble(2*l) )
								admit_error(l+1) = sqrt(admit_error(l+1))
							endif
							
						endif
						
					endif
					
				endif
				
				if (present(corr_error)) then
				
					if (K > 1) then 
					
						corr_error(l+1) = sum( ( corr_k(l+1,1:K) - corr(l+1) )**2 ) / dble(K-1) / dble(K) ! standard error!
						corr_error(l+1) = sqrt(corr_error(l+1))
					endif
					
				endif
				
			enddo
			
		endif
			
			
		deallocate(dj)
		deallocate(shwin)
		deallocate(shwinrot)
		deallocate(shloc_g)
		deallocate(shloc_t)
		
	endif	
	
end subroutine SHLocalizedAdmitCorr


